pm/floatmonitor.go

```diff
@@ -0,0 +1,132 @@
+package pm
+
+import (
+	"math/big"
+	"sync"
+	"time"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+// unixNow returns the current unix time
+// This is a wrapper function that can be stubbed in tests
+var unixNow = func() int64 {
+	return time.Now().Unix()
+}
+
+// FloatMonitor is an interface that describes methods used to
+// monitor a sender's max float which is based on the sender's available reserve
+type FloatMonitor interface {
+	// Add adds amount to the max float for addr
+	Add(addr ethcommon.Address, amount *big.Int) error
+	// Sub subtracts amount from the max float for addr
+	Sub(addr ethcommon.Address, amount *big.Int) error
+	// MaxFloat returns the current max float for addr
+	MaxFloat(addr ethcommon.Address) (*big.Int, error)
+}
+
+type item struct {
+	val        *big.Int
+	lastAccess int64
+}
+
+type floatMonitor struct {
+	claimant        ethcommon.Address
+	cleanupInterval time.Duration
+	ttl             int
+
+	mu        sync.Mutex
+	maxFloats map[ethcommon.Address]*item
+
+	broker Broker
+}
+
+// NewFloatMonitor returns a new FloatMonitor
+func NewFloatMonitor(claimant ethcommon.Address, broker Broker, cleanupInterval time.Duration, ttl int) FloatMonitor {
+	fm := &floatMonitor{
+		claimant:        claimant,
+		cleanupInterval: cleanupInterval,
+		ttl:             ttl,
+		broker:          broker,
+		maxFloats:       make(map[ethcommon.Address]*item),
+	}
+
+	ticker := time.NewTicker(fm.cleanupInterval)
+
+	// Start cleanup worker
+	// Every cleanupInterval the worker will remove itmes
+	// that have exceeded their ttl
+	go func() {
+		for {
+			<-ticker.C
+			fm.cleanup()
+		}
+	}()
+
+	return fm
+}
+
+func (fm *floatMonitor) Add(addr ethcommon.Address, amount *big.Int) error {
+	fm.mu.Lock()
+	defer fm.mu.Unlock()
+
+	maxFloat, err := fm.maxFloat(addr)
+	if err != nil {
+		return err
+	}
+
+	fm.maxFloats[addr].val = new(big.Int).Add(maxFloat, amount)
+
+	return nil
+}
+
+func (fm *floatMonitor) Sub(addr ethcommon.Address, amount *big.Int) error {
+	fm.mu.Lock()
+	defer fm.mu.Unlock()
+
+	maxFloat, err := fm.maxFloat(addr)
+	if err != nil {
+		return err
+	}
+
+	fm.maxFloats[addr].val = new(big.Int).Sub(maxFloat, amount)
+
+	return nil
+}
+
+func (fm *floatMonitor) MaxFloat(addr ethcommon.Address) (*big.Int, error) {
+	fm.mu.Lock()
+	defer fm.mu.Unlock()
+
+	return fm.maxFloat(addr)
+}
+
+func (fm *floatMonitor) maxFloat(addr ethcommon.Address) (*big.Int, error) {
+	// Request max float and cache it for addr if it
+	// is not already cached
+	if fm.maxFloats[addr] == nil {
+		maxFloat, err := fm.broker.ClaimableReserve(addr, fm.claimant)
+		if err != nil {
+			return nil, err
+		}
+
+		fm.maxFloats[addr] = &item{
+			val: maxFloat,
+		}
+	}
+
+	fm.maxFloats[addr].lastAccess = unixNow()
```

Comment:
Could we run into concurrency issues here without the use of a lock?

Reply:
We're using a lock in the exported wrapper function (`MaxFloat`) and defering unlocking it until `maxFloat` finishes (the helper). 

This is suboptimal however since we're locking the entire map and not the key/value pair for a certain address.

Reply:
While I initially considered a key/value pair lock, I just remembered that even with a key/value pair lock, the global map lock is unavoidable if you want to eliminate data races because the underlying implementation for a map can involve internal data updates (i.e. rebalancing to maintain even value distribution across buckets to ensure O(1) lookups in a hash map implementation - see: https://dave.cheney.net/2015/12/07/are-go-maps-sensitive-to-data-races) that impact the entirety of the map when updating a single key.

The one improvement that could be considered here is to use sync.RWMutex instead of sync.Mutex because the former will allow for multiple non-blocking concurrent readers - in *most* cases, `FloatMonitor` is probably going to be used for reading values and writing values will only occur when a winning ticket is being redeemed.

Reply:
Now using an [sync.RWMutex](https://github.com/livepeer/go-livepeer/pull/904/files#diff-2bfb8ebb26cc77da0bad94931ddc85f0R38).